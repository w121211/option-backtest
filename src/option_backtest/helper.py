from datetime import datetime


def generate_output_dir(
    config: dict,
    output_dir: str,
    include_params=["stock_id", "s0", "mu", "sigma_stock", "start_date", "end_date"],
    exclude_params: set[str] = set(),
) -> str:
    output_dir_parts = []

    for param in include_params:
        if param not in exclude_params:
            value = config.get(param)
            if isinstance(value, datetime):
                value = value.strftime("%Y%m%d")
            elif isinstance(value, float):
                value = str(int(value * 100))
            output_dir_parts.append(f"{param}_{value}")

    # output_dir = (
    #     output_dir
    #     # + datetime.now().strftime("%Y%m%d_%H%M%S_")
    #     + "_".join(output_dir_parts)
    # )
    output_dir = f"{output_dir}/{'_'.join(output_dir_parts)}"

    return output_dir
