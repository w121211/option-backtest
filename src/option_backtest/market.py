import copy
import logging
from dataclasses import dataclass
from typing import Literal, Protocol

import numpy as np
import pandas as pd
import plotly.graph_objects as go
from scipy.stats import norm


logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__name__)


@dataclass
class Stock:
    id: str
    df: pd.DataFrame  # Columns = ['price']. Indexed by 'date'
    position: int = 0
    # entry_date: datetime | None = None
    # exit_date: datetime | None = None

    def __repr__(self):
        return f"Stock(id='{self.id}', position={self.position})"


@dataclass
class Option:
    id: str
    option_type: Literal["call", "put"]
    strike: float
    expiry_date: pd.Timestamp
    df: pd.DataFrame  # Columns ['price', 'delta', 'vega', ...], index by 'date'

    position: int = 0
    # entry_date: datetime | None = None
    # exit_date: datetime | None = None

    def __repr__(self):
        return f"Option(id='{self.id}', position={self.position})"


@dataclass
class Account:
    cash: float
    stock: Stock  # Underlying stock
    options: dict[str, Option]  # key: option id, value: option


@dataclass
class Trade:
    asset_type: Literal["stock", "option"]
    asset_id: str
    amount: int
    date: pd.Timestamp

@dataclass
class MarketState:
    stock: Stock  # Underlying stock
    listed_options: dict[str, Option]
    account: Account
    current_date: pd.Timestamp
    current_step: int


class OnTradeFn(Protocol):
    def __call__(self, state: MarketState) -> list[Trade]: ...


def execute_trades(
    state: MarketState,
    trades: list[Trade],
) -> Account:
    listed_options = state.listed_options
    account = state.account
    current_date = state.current_date

    for trade in trades:
        if trade.asset_type == "stock":
            stock = account.stock
            stock_price = stock.df.at[current_date, "price"]
            stock.position += trade.amount
            account.cash -= stock_price * trade.amount

            logger.info(
                f"Executed trade [{current_date.strftime("%Y-%m-%d")}] {trade.asset_id} @ {stock_price:.2f} * {trade.amount}, Stock position: {stock.position}, Cash: {account.cash:.2f}"
            )

        elif trade.asset_type == "option":
            if trade.asset_id not in account.options:
                opt = copy.copy(listed_options[trade.asset_id])
                opt.position = 0
                account.options[opt.id] = opt

            opt = account.options[trade.asset_id]
            opt_price = opt.df.at[current_date, "price"]
            opt.position += trade.amount
            account.cash -= opt_price * trade.amount * 100

            logger.info(
                f"Executed trade[{current_date.strftime("%Y-%m-%d")}] {trade.asset_id} @ {opt_price:.2f} * {trade.amount}, Option position: {opt.position}, Cash: {account.cash:.2f}"
            )

    return account


def after_market(
    state: MarketState
) -> tuple[dict[str, Option], Account]:
    listed_options = state.listed_options
    current_date = state.current_date
    account = state.account

    # Remove expired options from available options
    listed_options = {
        opt_id: opt
        for opt_id, opt in listed_options.items()
        if opt.expiry_date > current_date
    }

    # Check account options
    for opt_id, opt in list(account.options.items()):
        if opt.expiry_date <= current_date:
            current_stock_price = account.stock.df.at[current_date, "price"]

            # Option has expired
            if opt.option_type == "call":
                if current_stock_price > opt.strike and opt.position != 0:
                    # Exercise call option
                    account.stock.position += opt.position * 100
                    account.cash -= opt.position * opt.strike * 100
                    logger.info(
                        f"Exercised call option {opt_id}. Bought {opt.position * 100} shares at strike price {opt.strike}. Current stock price={current_stock_price}"
                    )

            elif opt.option_type == "put":
                if current_stock_price < opt.strike and opt.position != 0:
                    # Exercise put option
                    account.stock.position -= opt.position * 100
                    account.cash += opt.position * opt.strike * 100
                    logger.info(
                        f"Exercised put option {opt_id}. Sold {opt.position * 100} shares at strike price {opt.strike}."
                    )

            # Remove expired option from account
            del account.options[opt_id]
            # logger.info(f"Removed expired option {opt_id} from account.")

    return listed_options, account


def calculate_portfolio_value(account: Account, current_date: pd.Timestamp) -> float:
    portfolio_value = account.cash

    # Calculate stock value
    stock_price = account.stock.df.at[current_date, "price"]
    stock_value = account.stock.position * stock_price
    portfolio_value += stock_value

    # Calculate option values
    for opt in account.options.values():
        if opt.expiry_date > current_date:
            opt_price = opt.df.at[current_date, "price"]
            opt_value = opt.position * opt_price * 100
            portfolio_value += opt_value

    return portfolio_value


def episode(
    init_cash: float,
    listed_stock: Stock,
    listed_options: dict[str, Option],
    on_trade_fns: list[OnTradeFn],  # A list of trading functions
):
    account = Account(cash=init_cash, stock=listed_stock, options={})

    log_trades = []
    log_portfolio_values = []

    for i, date in enumerate(account.stock.df.index):
        state = MarketState(
            stock=listed_stock,
            listed_options=listed_options,
            account=account,
            current_date=date,
            current_step=i,
        )

        trades = []
        for on_trade in on_trade_fns:
            _trades = on_trade(state)
            trades.extend(_trades)

        account = execute_trades(state, trades)
        listed_options, account = after_market(state)
        portfolio_value = calculate_portfolio_value(account, date)

        log_trades.extend(trades)
        log_portfolio_values.append(portfolio_value)

    portfolio_df = pd.DataFrame(
        {"value": log_portfolio_values}, index=account.stock.df.index
    )

    return log_trades, portfolio_df
