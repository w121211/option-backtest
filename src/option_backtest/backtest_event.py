# Enable inline dict typing. See https://github.com/microsoft/pyright/discussions/5682
# pyright: enableExperimentalFeatures=true

import copy
import os
import logging
from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Callable, Literal, TypedDict
from typing_extensions import NotRequired

import numpy as np
import pandas as pd
import plotly.graph_objects as go

from .gen_data import generate_and_save_stock_data, generate_options
from .market import Account, Option, Stock, StrategyFunc, Trade, episode
from .option import FindOptionsBy, find_options

logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__name__)


def prepare_market_df(stock: Stock, options: list[Option]) -> pd.DataFrame:
    market_df = pd.DataFrame(index=stock.df.index)
    market_df[f"{stock.id}_price"] = stock.df["price"]

    for option in options:
        market_df[f"{option.id}_price"] = option.df["price"]

    return market_df


# def monte_carlo_demo(
#     sample_price_series: Callable,
#     generate_options: Callable,
#     strategy: Callable,
#     n_trials: int,
# ):
#     trial_portfolio_dfs = []

#     for i in range(n_trials):
#         price_series = sample_price_series()
#         stock = Stock(ticker="XYZ", df=price_series)
#         options = generate_options(price_series)
#         market_df = prepare_market_df(stock, options)

#         portfolio_df, trades = episode(market_df, strategy)
#         trial_portfolio_dfs.append(portfolio_df)

#     # Concatenate trial portfolio dataframes
#     combined_portfolio_df = pd.concat(trial_portfolio_dfs)

#     # Calculate metrics across trials
#     metrics = {
#         "Total Return": combined_portfolio_df["portfolio_value"].pct_change().mean(),
#         "Volatility": combined_portfolio_df["portfolio_value"].pct_change().std(),
#         # Add more metrics as needed
#     }

#     return metrics


def calculate_metrics(portfolio_df: pd.DataFrame, risk_free_rate: float = 0.02) -> dict:
    returns = portfolio_df["value"].pct_change()
    total_return = (portfolio_df["value"].iloc[-1] / portfolio_df["value"].iloc[0]) - 1
    sharpe_ratio = (returns.mean() - risk_free_rate) / returns.std() * np.sqrt(252)
    max_drawdown = (portfolio_df["value"] / portfolio_df["value"].cummax() - 1).min()

    metrics = {
        "Total Return": total_return,
        "Sharpe Ratio": sharpe_ratio,
        "Max Drawdown": max_drawdown,
        "Volatility": returns.std() * np.sqrt(252),
    }
    return metrics


def plot_results(
    stock_df: pd.DataFrame,
    listed_options: dict[str, Option],
    portfolio_df: pd.DataFrame,
    trades: list[Trade],
):
    fig = go.Figure()

    fig.add_trace(
        go.Scatter(
            x=stock_df.index,
            y=stock_df["price"],
            name="Stock Price",
            line=dict(color="blue"),
            yaxis="y1",
        )
    )

    fig.add_trace(
        go.Scatter(
            x=portfolio_df.index,
            y=portfolio_df["value"],
            name="Portfolio Value",
            line=dict(color="green"),
            yaxis="y2",
        )
    )

    # Add entry trades as markers
    for trade in trades:
        if trade.asset_type == "option" and trade.position > 0:
            option = next(
                opt for opt in listed_options.values() if opt.id == trade.asset_id
            )
            fig.add_trace(
                go.Scatter(
                    x=[trade.date],
                    y=[option.strike],
                    mode="markers",
                    marker=dict(
                        symbol="triangle-up",
                        size=10,
                        color="red",
                        # line=dict(width=2, color="DarkSlateGrey"),
                    ),
                    name=f"Entry: {option.id}",
                    hovertext=f"Strike: {option.strike}<br>Expiry: {option.expiry_date.strftime('%y%m%d')}",
                    yaxis="y1",
                    legendgroup="markers",  # Set the same legendgroup for all markers
                    showlegend=False,  # Hide the markers from the legend
                )
            )
        if trade.asset_type == "option" and trade.position < 0:
            option = next(
                opt for opt in listed_options.values() if opt.id == trade.asset_id
            )
            fig.add_trace(
                go.Scatter(
                    x=[trade.date],
                    y=[option.strike],
                    mode="markers",
                    marker=dict(
                        symbol="triangle-down",
                        size=10,
                        color="darkred",
                        # line=dict(width=2, color="DarkSlateGrey"),
                    ),
                    name=f"Entry: {option.id}",
                    hovertext=f"Strike: {option.strike}<br>Expiry: {option.expiry_date.strftime('%y%m%d')}",
                    yaxis="y1",
                    legendgroup="markers",  # Set the same legendgroup for all markers
                    showlegend=False,  # Hide the markers from the legend
                )
            )

    fig.update_layout(
        title="Stock Price, Portfolio Value, and Entry Trades",
        # xaxis_title="Date",
        xaxis=dict(
            title="Date",
            tickformat="%y-%m-%d",  # Set the date format for the x-axis
        ),
        yaxis=dict(
            title="Stock Price",
            titlefont=dict(color="blue"),
            tickfont=dict(color="blue"),
        ),
        yaxis2=dict(
            title="Portfolio Value",
            titlefont=dict(color="green"),
            tickfont=dict(color="green"),
            overlaying="y",
            side="right",
        ),
        legend=dict(x=0, y=1, orientation="h"),
    )

    fig.show()


# -------- Strategy --------


def count_option_legs(account: Account) -> int:
    legs = 0
    for option in account.options.values():
        if option.position is not None and option.position != 0:
            legs += 1
    return legs


def get_current_option(account: Account) -> Option | None:
    for option in account.options.values():
        if option.position != 0:
            return option
    return None


ShouldRolloverCriteria = TypedDict(
    "ShouldRolloverCriteria",
    {
        "dte_lt": NotRequired[int],
        "delta_gt": NotRequired[float],
        "delta_lt": NotRequired[float],
    },
)


def should_rollover(
    current_date: datetime,
    current_option: Option | None,
    criteria: ShouldRolloverCriteria,
) -> bool:
    # Example rollover criteria:
    # - If the current option has less than 10 days to expiry
    # - If the current option's delta is less than 0.3 or greater than 0.7
    if current_option is None:
        return True

    if criteria.get("dte_lt"):
        dte = (current_option.expiry_date - current_date).days
        return dte < criteria["dte_lt"]  # type: ignore

    return False

    # return (
    #     days_to_expiry < 10
    #     or current_option.df.loc[current_date, "delta"] < 0.3
    #     or current_option.df.loc[current_date, "delta"] > 0.7
    # )


def create_rollover_strategy(
    n_option_legs: int = 1,
    entry_position: int = -1,
    find_options_by: FindOptionsBy | None = None,
    rollover_criteria: ShouldRolloverCriteria | None = None,
) -> StrategyFunc:
    if find_options_by is None:
        find_options_by = {
            "option_type": "put",
            "dte_gt": 29,
            "dte_lt": 61,
            "sort_by_delta_near": -0.5,
        }
    if rollover_criteria is None:
        rollover_criteria = {"dte_lt": 2}

    def rollover_strategy(
        listed_options: dict[str, Option],
        account: Account,
        current_date: pd.Timestamp,
    ) -> list[Trade]:
        # Validation
        if count_option_legs(account) > n_option_legs:
            raise Exception("This strategy is one-leg only.")

        current_option = get_current_option(account)
        if should_rollover(current_date, current_option, criteria=rollover_criteria):
            current_stock_price = account.stock.df.at[current_date, "price"]
            found_options = find_options(
                listed_options, current_stock_price, current_date, find_options_by
            )
            if not found_options:
                logger.warning(
                    f"No suitable options found for rollover on {current_date}"
                )
                return []

            next_option = found_options[0]

            trades: list[Trade] = []

            # Exit the current option position
            if current_option is not None:
                trades.append(
                    Trade(
                        asset_type="option",
                        asset_id=current_option.id,
                        position=-current_option.position,
                        date=current_date,
                    )
                )

            # Enter the new option position
            trades.append(
                Trade(
                    asset_type="option",
                    asset_id=next_option.id,
                    position=entry_position,
                    date=current_date,
                )
            )

            next_option_delta = next_option.df.at[current_date, "delta"]
            logger.info(
                f"Rollover date={current_date.strftime('%y%m%d')}, price={current_stock_price:.2f}, delta={next_option_delta:.2f}, {current_option.id if current_option else 'None'} -> {next_option.id}"
            )
            return trades
        else:
            logger.info(f"No rollover needed on {current_date}")
            return []

    return rollover_strategy


def demo(strategy_fn: StrategyFunc):
    # Read or generate stock data
    stock_id = "XYZ"
    output_dir = f"output/{stock_id}"

    s0 = 100.0  # 初始價格
    mu = 0.05  # 預期收益率（年化）
    sigma = 0.2  # 波動率（年化）
    dt = 1 / 252  # 時間步長（假設一年有252個交易日）
    n_steps = 252  # 模擬步數（假設模擬一年的價格路徑）
    start_date = datetime(2023, 1, 1)
    end_date = datetime(2023, 12, 31)

    stock_price_path = f"{output_dir}/{stock_id}_price.csv"
    if not os.path.exists(stock_price_path):
        generate_and_save_stock_data(
            s0, mu, sigma, dt, start_date, end_date, stock_price_path
        )

    stock_price_df = pd.read_csv(stock_price_path)
    stock_price_df["date"] = pd.to_datetime(stock_price_df["date"])
    stock_price_df = stock_price_df.set_index("date")
    stock = Stock(id="XYZ", df=stock_price_df)

    # Generate options data
    r = 0.05  # Risk-free rate
    sigma_option = 0.2  # Volatility

    strikes = np.arange(
        (min(stock.df["price"]) // 5 - 2) * 5,
        (max(stock.df["price"]) // 5 + 3) * 5 + 1,
        5,
    )
    expiry_dates = list(pd.date_range("2023-01-01", periods=13, freq="M"))
    listed_options = generate_options(
        stock,
        strikes,
        expiry_dates,
        r,
        sigma_option,
    )

    # Run backtest
    init_cash = 1e3
    trades, portfolio_df = episode(init_cash, stock, listed_options, strategy_fn)

    # Calculate metrics
    metrics = calculate_metrics(portfolio_df)

    print("Metrics:")
    for metric, value in metrics.items():
        print(f"{metric}: {value:.4f}")

    # Plot results
    plot_results(stock.df, listed_options, portfolio_df, trades)


if __name__ == "__main__":
    # Long call, rollover before 2 days, roll to ATM, 30 days+
    n_option_legs = 1
    entry_position = 1
    find_options_by: FindOptionsBy = {
        "option_type": "call",
        "dte_gt": 29,
        "dte_lt": 61,
        "sort_by_delta_near": 0.5,
    }
    rollover_criteria: ShouldRolloverCriteria = {"dte_lt": 2}

    strategy = create_rollover_strategy(
        n_option_legs,
        entry_position,
        find_options_by,
        rollover_criteria,
    )
    demo(strategy)
