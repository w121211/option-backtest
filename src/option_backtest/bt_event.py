# Enable inline dict typing. See https://github.com/microsoft/pyright/discussions/5682
# pyright: enableExperimentalFeatures=true

import copy
import os
import logging
from dataclasses import asdict, dataclass, replace
from datetime import datetime, timedelta
from typing import Callable, Literal, TypedDict
from typing_extensions import NotRequired

import numpy as np
import pandas as pd
import plotly.graph_objects as go

from .gen_data import generate_stock_data, generate_options
from .market import Account, Option, Stock, Trade, OnTradeFn, episode


logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__name__)


@dataclass
class GenerateStockConfig:
    # Generate stock prices
    s0: float = 100.0  # 初始價格
    mu: float = 0.05  # 預期收益率（年化）
    sigma_stock: float = 0.2  # 波動率（年化）
    dt: float = 1 / 252  # 時間步長（假設一年有252個交易日）
    n_steps: int = 252  # 模擬步數（假設模擬一年的價格路徑）


@dataclass
class GenerateOptionsConfig:
    # Generate option prices
    r: float = 0.05  # Risk-free rate
    sigma_option: float = 0.2  # Volatility


@dataclass
class BacktestConfig:
    # Basic
    stock_id: str = "XYZ"  # Stock ticker
    output_dir: str = "./output"  # Backtest output directory
    stock_price_dir: str | None = None  # Stock price directory for monte carlo backtest
    stock_price_path: str | None = None  # Stock price for backtest

    # Required params
    start_date: datetime = datetime(2023, 1, 1)
    end_date: datetime = datetime(2023, 12, 31)
    r: float = 0.05  # Risk-free rate for calculating Sharpe Ratio

    # Episode
    init_cash: float = 1e3

    # Configs for generating data
    gen_stock_config: GenerateStockConfig | None = None
    gen_options_config: GenerateOptionsConfig | None = None


def prepare_market_df(stock: Stock, options: list[Option]) -> pd.DataFrame:
    market_df = pd.DataFrame(index=stock.df.index)
    market_df[f"{stock.id}_price"] = stock.df["price"]

    for option in options:
        market_df[f"{option.id}_price"] = option.df["price"]

    return market_df


# --------- Process Backtest Result ---------


def calculate_metrics(
    portfolio_df: pd.DataFrame, risk_free_rate: float
) -> pd.DataFrame:
    returns = portfolio_df["value"].pct_change()  # 計算每日收益率
    total_return = (
        portfolio_df["value"].iloc[-1] / portfolio_df["value"].iloc[0]
    ) - 1  # 總回報率
    daily_risk_free_rate = risk_free_rate / 252  # 將年化無風險利率轉換為每日利率
    sharpe_ratio = (
        (returns.mean() - daily_risk_free_rate) / returns.std() * np.sqrt(252)
    )  # 計算年化Sharpe比率
    max_drawdown = (
        portfolio_df["value"] / portfolio_df["value"].cummax() - 1
    ).min()  # 最大回撤
    volatility = returns.std() * np.sqrt(252)  # 年化波動性
    std_dev = returns.std()  # 日波動性（標準差）

    metrics_data = {
        "Return": [total_return],
        "Sharpe": [sharpe_ratio],
        "MaxDrawdown": [max_drawdown],
        "Volatility": [volatility],
        "StdDev": [std_dev],  # 新增每日標準差
    }
    metrics_df = pd.DataFrame(metrics_data)
    return metrics_df


def save_outputs(
    config: BacktestConfig,
    trades: list[Trade],
    portfolio_df: pd.DataFrame,
    metrics_df: pd.DataFrame,
):
    output_dir = config.output_dir
    trades_path = f"{output_dir}/trades.csv"
    portfolio_path = f"{output_dir}/portfolio.csv"
    metrics_path = f"{output_dir}/metrics.csv"

    pd.DataFrame(trades).to_csv(trades_path, index=False)
    portfolio_df.to_csv(portfolio_path)
    metrics_df.to_csv(metrics_path, index=False)


def plot_results(
    stock_df: pd.DataFrame,
    listed_options: dict[str, Option],
    portfolio_df: pd.DataFrame,
    trades: list[Trade],
):
    fig = go.Figure()

    fig.add_trace(
        go.Scatter(
            x=stock_df.index,
            y=stock_df["price"],
            name="Stock Price",
            line=dict(color="blue"),
            yaxis="y1",
        )
    )

    fig.add_trace(
        go.Scatter(
            x=portfolio_df.index,
            y=portfolio_df["value"],
            name="Portfolio Value",
            line=dict(color="green"),
            yaxis="y2",
        )
    )

    # Add entry trades as markers
    for trade in trades:
        if trade.asset_type == "option" and trade.amount > 0:
            option = next(
                opt for opt in listed_options.values() if opt.id == trade.asset_id
            )
            fig.add_trace(
                go.Scatter(
                    x=[trade.date],
                    y=[option.strike],
                    mode="markers",
                    marker=dict(
                        symbol="triangle-up",
                        size=10,
                        color="red",
                        # line=dict(width=2, color="DarkSlateGrey"),
                    ),
                    name=f"Entry: {option.id}",
                    hovertext=f"Strike: {option.strike}<br>Expiry: {option.expiry_date.strftime('%y%m%d')}",
                    yaxis="y1",
                    legendgroup="markers",  # Set the same legendgroup for all markers
                    showlegend=False,  # Hide the markers from the legend
                )
            )
        if trade.asset_type == "option" and trade.amount < 0:
            option = next(
                opt for opt in listed_options.values() if opt.id == trade.asset_id
            )
            fig.add_trace(
                go.Scatter(
                    x=[trade.date],
                    y=[option.strike],
                    mode="markers",
                    marker=dict(
                        symbol="triangle-down",
                        size=10,
                        color="darkred",
                        # line=dict(width=2, color="DarkSlateGrey"),
                    ),
                    name=f"Entry: {option.id}",
                    hovertext=f"Strike: {option.strike}<br>Expiry: {option.expiry_date.strftime('%y%m%d')}",
                    yaxis="y1",
                    legendgroup="markers",  # Set the same legendgroup for all markers
                    showlegend=False,  # Hide the markers from the legend
                )
            )

    fig.update_layout(
        title="Stock Price, Portfolio Value, and Entry Trades",
        # xaxis_title="Date",
        xaxis=dict(
            title="Date",
            tickformat="%y-%m-%d",  # Set the date format for the x-axis
        ),
        yaxis=dict(
            title="Stock Price",
            titlefont=dict(color="blue"),
            tickfont=dict(color="blue"),
        ),
        yaxis2=dict(
            title="Portfolio Value",
            titlefont=dict(color="green"),
            tickfont=dict(color="green"),
            overlaying="y",
            side="right",
        ),
        legend=dict(x=0, y=1, orientation="h"),
    )

    fig.show()


# --------- Backtest ---------


def run_backtest(
    config: BacktestConfig,
    on_trade_fns: list[OnTradeFn],
    stock_price_df: pd.DataFrame | None = None,
    listed_options: dict[str, Option] | None = None,
) -> tuple[
    list[Trade],
    pd.DataFrame,
    pd.DataFrame,
    dict[str, Option],
]:
    if stock_price_df is None:
        # Read stock price data from the path
        stock_price_df = pd.read_csv(config.stock_price_path)
        stock_price_df["date"] = pd.to_datetime(stock_price_df["date"])
        stock_price_df = stock_price_df.set_index("date")

    stock = Stock(id="XYZ", df=stock_price_df)

    if listed_options is None:
        if config.gen_options_config is not None:
            # Generate options data based on stock prices
            r = config.gen_options_config.r  # Risk-free rate
            sigma_option = config.gen_options_config.sigma_option  # Volatility

            strikes = np.arange(
                (min(stock.df["price"]) // 5 - 2) * 5,
                (max(stock.df["price"]) // 5 + 3) * 5 + 1,
                5,
            )
            expiry_dates = list(pd.date_range("2023-01-01", periods=104, freq="W"))
            listed_options = generate_options(
                stock,
                strikes,
                expiry_dates,
                r,
                sigma_option,
            )
        else:
            raise Exception("config.gen_options_config is None")

    # Run backtest
    trades, portfolio_df = episode(
        config.init_cash, stock, listed_options, on_trade_fns
    )

    return trades, portfolio_df, stock_price_df, listed_options


def run_monte_carlo_backtest(
    config: BacktestConfig,
    on_trade_fns: list[OnTradeFn],
    num_simulations: int,
) -> pd.DataFrame:
    """
    Run Monte Carlo simulations of the backtest.

    Parameters:
    - config: BacktestConfig object containing the backtest configuration.
    - strategies: List of trading strategies to be used in the backtest.
    - num_simulations: Number of Monte Carlo simulations to run.

    Returns:
    - A DataFrame containing the performance metrics for each simulation.
    """

    def get_stock_price_path(sim_index: int) -> str:
        return f"./{config.stock_price_dir}/{config.stock_id}_price_sim_{sim_index}.csv"

    performance_metrics = []

    # Generate stock data if not existed
    for i in range(num_simulations):
        stock_price_path = get_stock_price_path(i)
        if not os.path.exists(stock_price_path) and config.gen_stock_config is not None:
            generate_stock_data(
                s0=config.gen_stock_config.s0,
                mu=config.gen_stock_config.mu,
                sigma=config.gen_stock_config.sigma_stock,
                dt=config.gen_stock_config.dt,
                start_date=config.start_date,
                end_date=config.end_date,
                output_path=stock_price_path,
            )
        else:
            raise Exception("no stock_price_path & config.gen_stock_config is None")

    for i in range(num_simulations):
        print(f"Running simulation {i+1}/{num_simulations}")

        # Update the backtest configuration with the random parameters
        config_sim = copy.deepcopy(config)
        config_sim.stock_price_path = get_stock_price_path(i)

        # Create a unique output directory for each simulation
        config_sim.output_dir = f"{config.output_dir}/sim_{i+1}"
        os.makedirs(config_sim.output_dir, exist_ok=True)

        # Run backtest
        trades, portfolio_df, stock_price_df, listed_options = run_backtest(
            config_sim, on_trade_fns
        )

        # Extract performance metrics from backtest results
        metrics = calculate_metrics(portfolio_df, config.r)
        metrics["Simulation"] = i + 1
        performance_metrics.append(metrics)

    performance_df = pd.DataFrame(performance_metrics)

    # Save the performance metrics dataframe
    performance_path = f"{config.output_dir}/performance_metrics.csv"
    performance_df.to_csv(performance_path, index=False)

    return performance_df
