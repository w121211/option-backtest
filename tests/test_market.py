from datetime import datetime, timedelta
import pytest
import pandas as pd

from src.option_backtest.market import (
    Stock,
    Option,
    Account,
    Trade,
    after_market,
    execute_trades,
)


@pytest.fixture
def current_date():
    return pd.Timestamp("2023-01-01")


@pytest.fixture
def listed_stock(current_date):
    dates = pd.date_range(start=current_date, periods=2, freq="D")
    df = pd.DataFrame({"price": [100, 105]}, index=dates)
    return Stock(id="XYZ", df=df)


@pytest.fixture
def listed_options(current_date):
    expiry_date1 = current_date + timedelta(days=10)
    expiry_date2 = current_date - timedelta(days=1)
    expiry_date3 = current_date

    options = [
        Option(
            id="XYZ_C_110_10",
            option_type="call",
            strike=110,
            expiry_date=expiry_date1,
            df=pd.DataFrame(
                {"price": [5, 6]},
                index=pd.date_range(start=current_date, periods=2, freq="D"),
            ),
        ),
        Option(
            id="XYZ_P_90_m1",
            option_type="put",
            strike=90,
            expiry_date=expiry_date2,
            df=pd.DataFrame(
                {"price": [3, 4]},
                index=pd.date_range(start=current_date, periods=2, freq="D"),
            ),
        ),
        Option(
            id="XYZ_C_90_0",
            option_type="call",
            strike=90,
            expiry_date=expiry_date3,
            df=pd.DataFrame(
                {"price": [3, 4]},
                index=pd.date_range(start=current_date, periods=2, freq="D"),
            ),
        ),
    ]
    return {opt.id: opt for opt in options}


def test_execute_trades(current_date, listed_stock, listed_options, snapshot):
    _options = list(listed_options.values())
    account = Account(cash=1000, stock=listed_stock, options={})

    trades = [
        Trade(
            asset_type="option",
            asset_id=_options[0].id,
            position=2,
            date=current_date,
        ),
        Trade(
            asset_type="option",
            asset_id=_options[1].id,
            position=-1,
            date=current_date,
        ),
    ]
    account1 = execute_trades(listed_options, account, current_date, trades)
    assert account1 == snapshot

    trades = [
        Trade(asset_type="stock", asset_id="XYZ", position=10, date=current_date),
    ]
    account2 = execute_trades(listed_options, account1, current_date, trades)
    assert account2 == snapshot


def test_after_market_expired_call_exercised(
    current_date, listed_stock, listed_options, snapshot
):
    account = Account(cash=1000, stock=listed_stock, options=listed_options)
    account.options["XYZ_C_110_10"].position = 1
    account.options["XYZ_P_90_m1"].position = 1
    account.options["XYZ_C_90_0"].position = 1

    listed_options, updated_account = after_market(
        current_date, listed_options, account
    )

    # assert len(available_options) == 1
    # assert "XYZ_C_110_20230112" in available_options
    # assert "XYZ_P_90_20230101" not in available_options

    # assert len(updated_account.options) == 1
    # assert "XYZ_C_110_20230112" in updated_account.options
    # assert "XYZ_P_90_20230101" not in updated_account.options

    # assert updated_account.stock.position == 0
    # assert updated_account.cash == 1000

    assert listed_options == snapshot
    assert updated_account == snapshot


# def test_after_market_expired_put_exercised(current_date, listed_stock, listed_options):
#     listed_stock.df.at[current_date, "price"] = 80
#     account = Account(cash=1000, stock=listed_stock, options=listed_options)
#     account.stock.position = 1
#     account.options["XYZ_C_110_20230112"].position = 1
#     account.options["XYZ_P_90_20230101"].position = 1

#     available_options, updated_account = after_market(
#         current_date, listed_options, account
#     )

#     assert len(available_options) == 1
#     assert "XYZ_C_110_20230112" in available_options
#     assert "XYZ_P_90_20230101" not in available_options

#     assert len(updated_account.options) == 1
#     assert "XYZ_C_110_20230112" in updated_account.options
#     assert "XYZ_P_90_20230101" not in updated_account.options

#     assert updated_account.stock.position == 0
#     assert updated_account.cash == 1090
