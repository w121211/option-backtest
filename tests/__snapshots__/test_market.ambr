# serializer version: 1
# name: test_after_market_expired_call_exercised
  dict({
    'XYZ_C_110_10': Option(id='XYZ_C_110_10', position=1),
  })
# ---
# name: test_after_market_expired_call_exercised.1
  Account(cash=-8000, stock=Stock(id='XYZ', position=100), options={'XYZ_C_110_10': Option(id='XYZ_C_110_10', position=1)})
# ---
# name: test_execute_trades
  Account(cash=993, stock=Stock(id='XYZ', position=0), options={'XYZ_C_110_10': Option(id='XYZ_C_110_10', position=2), 'XYZ_P_90_m1': Option(id='XYZ_P_90_m1', position=-1)})
# ---
# name: test_execute_trades.1
  Account(cash=-7, stock=Stock(id='XYZ', position=10), options={'XYZ_C_110_10': Option(id='XYZ_C_110_10', position=2), 'XYZ_P_90_m1': Option(id='XYZ_P_90_m1', position=-1)})
# ---
